File: 974-subarray-sums-divisible-by-k/974-subarray-sums-divisible-by-k.cpp
class Solution {
public:
    int subarraysDivByK(vector<int>& nums, int k) {
        vector<int>count(k,0);
        count[0]=1;
        int ans=0;
        int prefix=0;
        for(int it:nums){
            prefix+=(it%k+k)%k;
            ans+=count[prefix%k]++;
        }
        return ans;
    }
};

File: 974-subarray-sums-divisible-by-k/974-subarray-sums-divisible-by-k.py
from collections import defaultdict
class Solution:
    def subarraysDivByK(self, nums: List[int], k: int) -> int:
        freq = [0]*k
        prefix, rem, count = 0, 0, 0
        freq[0] = 1
        for i in nums:
            prefix += i
            rem = prefix % k
            if rem < 0:
                rem += k
            count += freq[rem]
            freq[rem] += 1
        
        return count